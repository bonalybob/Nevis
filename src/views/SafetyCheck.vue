<template>
  <main>
    <back-arrow :to="`/events/${$route.params.id}`" />
    <h1 class="mx-10 mb-1 px-1">
      Safety Check
    </h1>
    <div class="mx-12 mb-3">
      <button class="button" @click="importPunches">Import Punches</button>
      <button class="button" @click="getOutstandingCompetitors">Refresh</button>
    </div>
    <div
      v-if="outstandingCompetitors && outstandingCompetitors.length > 0"
      class="mx-12 my-shadow mx-12 mb-3 p-2"
    >
      <table class="w-full font-body">
        <tr class="font-heading text-center hover:bg-blue-light">
          <th @click="changeSortBy('name')">
            Name
            <up-down-arrows
              :active="sortBy === 'name'"
              :ascending="sortDirection === 'ASC'"
            />
          </th>
          <th @click="changeSortBy('controlCode')">
            Control
            <up-down-arrows
              :active="sortBy === 'controlCode'"
              :ascending="sortDirection === 'ASC'"
            />
          </th>
          <th @click="changeSortBy('time')">
            Time
            <up-down-arrows
              :active="sortBy === 'time'"
              :ascending="sortDirection === 'ASC'"
            />
          </th>
          <th
            class="hidden sm:table-cell"
            @click="changeSortBy('competitors.course')"
          >
            Course
            <up-down-arrows
              :active="sortBy === 'competitors.course'"
              :ascending="sortDirection === 'ASC'"
            />
          </th>
          <th class="hidden md:table-cell" @click="changeSortBy('ageClass')">
            Age Class
            <up-down-arrows
              :active="sortBy === 'ageClass'"
              :ascending="sortDirection === 'ASC'"
            />
          </th>
        </tr>
        <tbody is="transition-group" name="fade">
          <tr
            v-for="competitor of outstandingCompetitors"
            :key="competitor.punchesId"
            class="text-center even:bg-blue-lightest hover:bg-blue-light"
          >
            <td>{{ competitor.name }}</td>
            <td>{{ competitor.controlCode }}</td>
            <td>{{ time.displayActualTime(competitor.time) }}</td>
            <td class="hidden sm:table-cell">{{ competitor.course }}</td>
            <td class="hidden md:table-cell">{{ competitor.ageClass }}</td>
          </tr>
        </tbody>
      </table>
    </div>
  </main>
</template>

<script>
import BackArrow from '@/components/BackArrow'
import UpDownArrows from '@/components/UpDownArrows'

import time from '@/scripts/time'
import ageClassFunctions from '@/scripts/ageClass'

export default {
  components: {
    'back-arrow': BackArrow,
    'up-down-arrows': UpDownArrows,
  },

  data: function () {
    return {
      outstandingCompetitors: [],
      time: time,
      sortBy: 'time',
      sortDirection: 'DESC',
    }
  },

  mounted: function () {
    if (this.$database.connection === null || !this.$database.connected) {
      this.$router.push('/')
      this.$messages.addMessage('Problem Connecting To Database', 'error')
    } else this.getOutstandingCompetitors()
  },

  methods: {
    changeSortBy: function (field) {
      if (this.sortBy === field && this.sortDirection === 'ASC')
        this.sortDirection = 'DESC'
      else if (this.sortBy === field) this.sortDirection = 'ASC'
      this.sortBy = field
      this.getOutstandingCompetitors()
    },

    getOutstandingCompetitors: function () {
      return this.$database
        .query(
          `
          SELECT time, controlCode, punches.id as punchesId, competitors.*, courses.name as course
          FROM punches, competitors
          LEFT JOIN courses ON competitors.course=courses.id
          WHERE punches.competitor=competitors.id
              AND competitors.downloaded=False
			        AND punches.event=?
          GROUP BY competitors.id
          ORDER BY ${this.sortBy} ${this.sortDirection}
      `,
          [this.$route.params.id]
        )
        .then((result) => {
          this.outstandingCompetitors = result
        })
        .catch(() =>
          this.$messages.addMessage(
            'Problem Fetching Outstanding Competitors',
            'error'
          )
        )
    },

    rawFileToPunchRecords: function (rawData) {
      return rawData
        .trim()
        .split('\n')
        .filter((record) => record.slice(0, 3) !== 'No;' && record !== '')
        .map((line) => {
          let splitLine = line.split(';')
          return {
            siid: parseInt(splitLine[2], 10),
            controlCode: parseInt(splitLine[6], 10),
            time: time.timeToSeconds(splitLine[8]),
            event: this.$route.params.id,
            type: 'imported',
          }
        })
    },

    findCompetitorForPunches: async function (punchRecords) {
      const results = []

      for (const record of punchRecords) {
        const competitorsMatchingSIID = await this.$database.query(
          `
          SELECT *
          FROM competitors
          WHERE competitors.event=?
            AND competitors.siid=?
          ORDER BY downloaded DESC`,
          [this.$route.params.id, record.siid.toString()]
        )

        if (competitorsMatchingSIID.length >= 1)
          results.push({ ...record, competitor: competitorsMatchingSIID[0].id })
        else {
          const archiveCompetitors = await this.$archive.query(
            'SELECT * FROM people WHERE siid=? AND status != "Hire" LIMIT 1',
            record.siid
          )

          let newCompetitor = {}
          if (archiveCompetitors.length >= 1)
            newCompetitor = {
              name: archiveCompetitors[0].name,
              event: this.$route.params.id,
              siid: record.siid,
              membershipNumber: archiveCompetitors[0].membershipNumber,
              ageClass: ageClassFunctions(
                archiveCompetitors[0].gender,
                archiveCompetitors[0].yearOfBirth
              ),
              club: archiveCompetitors[0].club,
              downloaded: false,
            }
          else
            newCompetitor = {
              name: 'Unknown',
              event: this.$route.params.id,
              siid: record.siid,
              downloaded: false,
            }

          const result = await this.$database.query(
            'INSERT INTO competitors SET ?',
            newCompetitor
          )

          results.push({ ...record, competitor: result.insertId })
        }
      }

      return results
    },

    punchRecordToArray: function (punchRecords) {
      return punchRecords.map((record) => [
        record.competitor,
        record.controlCode,
        record.time,
        record.event,
        record.type,
      ])
    },

    filterExistingPunches: async function (punchRecords) {
      const existingPunches = await this.$database.query(
        `
          SELECT siid, time, controlCode
          FROM punches, competitors
          WHERE punches.event=?
              AND competitors.id=punches.competitor
           `,
        [this.$route.params.id]
      )

      return punchRecords.filter((record) => {
        const punch = [record.siid, record.time, record.controlCode]
        for (const exisitingPunch of existingPunches) {
          if (JSON.stringify(exisitingPunch) === JSON.stringify(punch))
            return false
        }
        return true
      })
    },

    importPunches: function () {
      const { dialog } = this.$electron.remote
      return dialog
        .showOpenDialog({
          title: 'Nevis - Import Punches',
          buttonLabel: 'Import',
          properties: ['openFile'],
          filters: [
            { name: 'Comma Separated Values', extensions: ['csv'] },
            { name: 'All Files', extensions: ['*'] },
          ],
        })
        .then((result) => {
          if (!result.canceled)
            return this.$fs.readFile(result.filePaths[0], { encoding: 'utf8' })
          else throw Error()
        })
        .then(this.rawFileToPunchRecords)
        .then(this.findCompetitorForPunches)
        .then(this.filterExistingPunches)
        .then(this.punchRecordToArray)
        .then((result) => {
          if (result.length > 0)
            return this.$database.query(
              'INSERT INTO punches (competitor,controlCode,time,event,type) VALUES ?',
              [result]
            )
          else return { affectedRows: 0 }
        })
        .then((result) =>
          this.$messages.addMessage(`${result.affectedRows} Punches Imported`)
        )
        .then(this.getOutstandingCompetitors)
        .catch(() =>
          this.$messages.addMessage('Problem Importing Punches', 'error')
        )
    },
  },
}
</script>
