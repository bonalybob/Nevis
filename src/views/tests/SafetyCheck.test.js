import { shallowMount } from '@vue/test-utils'
import SafetyCheck from '@/views/SafetyCheck'

test('Is a Vue Instance', () => {
  const wrapper = shallowMount(SafetyCheck, {
    stubs: ['router-link'],
    mocks: {
      $database: {
        connection: {},
        connected: true,
        query: jest.fn().mockRejectedValue(),
      },
      $route: { params: { eventId: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
    },
  })
  expect(wrapper.isVueInstance()).toBeTruthy()
})

test('Renders Correctly', () => {
  const wrapper = shallowMount(SafetyCheck, {
    stubs: ['router-link'],
    mocks: {
      $archive: {
        connection: {},
        connected: true,
        query: jest.fn().mockRejectedValue(),
      },
      $database: {
        connection: {},
        connected: true,
        query: jest.fn().mockRejectedValue(),
      },
      $route: { params: { eventId: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
    },
  })
  wrapper.setData({
    outstandingCompetitors: [
      { name: 'Bob', controlCode: 202, time: 5655, course: 'Long' },
      { name: 'James', controlCode: 202, time: 4555, course: 'Short' },
      { name: 'Alfie', controlCode: 201, time: 5665, club: 'HAT' },
    ],
  })
  expect(wrapper.element).toMatchSnapshot()
})

test('Not Connected to the Database', () => {
  const wrapper = shallowMount(SafetyCheck, {
    stubs: ['router-link'],
    mocks: {
      $database: {
        connection: {},
        connected: false,
        query: jest.fn().mockRejectedValue(),
      },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
      $serialPort: {
        list: jest.fn().mockResolvedValue([]),
        open: jest.fn(),
        close: jest.fn(),
      },
    },
  })
  expect(wrapper.vm.$router.push).toHaveBeenCalledTimes(1)
  expect(wrapper.vm.$router.push).toHaveBeenLastCalledWith('/')
  expect(wrapper.vm.$messages.addMessage).toHaveBeenLastCalledWith(
    'Problem Connecting To Database',
    'error'
  )
})

test('Punch Record To Array', () => {
  const wrapper = shallowMount(SafetyCheck, {
    stubs: ['router-link'],
    mocks: {
      $archive: {
        connection: {},
        connected: true,
        query: jest.fn().mockRejectedValue(),
      },
      $database: {
        connection: {},
        connected: true,
        query: jest.fn().mockRejectedValue(),
      },
      $route: { params: { eventId: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
    },
  })
  expect(wrapper.vm.punchRecordToArray([])).toEqual([])
  expect(wrapper.vm.punchRecordToArray([{}])).toEqual([
    [undefined, undefined, undefined, undefined, undefined],
  ])
  expect(
    wrapper.vm.punchRecordToArray([
      {
        competitor: 7,
        controlCode: 202,
        time: 777,
        event: 3,
        type: 'imported',
      },
      {
        competitor: 5,
        controlCode: 202,
        time: 547,
        event: 3,
        type: 'imported',
      },
    ])
  ).toEqual([
    [7, 202, 777, 3, 'imported'],
    [5, 202, 547, 3, 'imported'],
  ])
})

test('Raw File to Punch Records', () => {
  const wrapper = shallowMount(SafetyCheck, {
    stubs: ['router-link'],
    mocks: {
      $archive: {
        connection: {},
        connected: true,
        query: jest.fn().mockRejectedValue(),
      },
      $database: {
        connection: {},
        connected: true,
        query: jest.fn().mockRejectedValue(),
      },
      $route: { params: { eventId: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
    },
  })
  expect(wrapper.vm.rawFileToPunchRecords(``)).toEqual([])
  expect(
    wrapper.vm.rawFileToPunchRecords(
      `No;Read on;SIID;Control time;Battery voltage;Serial number;Code number;DayOfWeek;Punch DateTime;Operating mode;SIAC number;SIAC Count;SIAC radio mode;SIAC is battery low;SIAC is card full;SIAC beacon mode;SIAC is gate mode;`
    )
  ).toEqual([])
  expect(
    wrapper.vm.rawFileToPunchRecords(
      `
      No;Read on;SIID;Control time;Battery voltage;Serial number;Code number;DayOfWeek;Punch DateTime;Operating mode;SIAC number;SIAC Count;SIAC radio mode;SIAC is battery low;SIAC is card full;SIAC beacon mode;SIAC is gate mode;
      1;2020-04-01 15:09:58;2047481;01/04/2020   13:59:22.468;;;102;We;13:59:22.468;Control;0;1;;;;;;
      2;2020-04-01 15:09:58;2047471;01/04/2020   13:59:26.312;;;102;We;13:59:26.312;Control;0;1;;;;;;
      3;2020-04-01 15:09:58;2047461;01/04/2020   13:59:29.148;;;102;We;13:59:29.148;Control;0;1;;;;;;
      4;2020-04-01 15:09:58;2047481;01/04/2020   13:59:40.398;;;102;We;13:59:40.398;Control;0;1;;;;;;

      `
    )
  ).toEqual([
    {
      controlCode: 102,
      event: undefined,
      siid: 2047481,
      time: 50362,
      type: 'imported',
    },
    {
      controlCode: 102,
      event: undefined,
      siid: 2047471,
      time: 50366,
      type: 'imported',
    },
    {
      controlCode: 102,
      event: undefined,
      siid: 2047461,
      time: 50369,
      type: 'imported',
    },
    {
      controlCode: 102,
      event: undefined,
      siid: 2047481,
      time: 50380,
      type: 'imported',
    },
  ])
})

test('Get Outstanding Competitors - Success', async () => {
  const wrapper = shallowMount(SafetyCheck, {
    stubs: ['router-link'],
    mocks: {
      $archive: {
        connection: {},
        connected: true,
        query: jest.fn().mockRejectedValue(),
      },
      $database: {
        connection: {},
        connected: true,
        query: jest.fn().mockResolvedValue([1, 2, 3]),
      },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
    },
  })
  await wrapper.vm.getOutstandingCompetitors()
  expect(wrapper.vm.$database.query).toHaveBeenCalledWith(expect.any(String), [
    12,
  ])
  expect(wrapper.vm.outstandingCompetitors).toEqual([1, 2, 3])
})
test('Get Outstanding Competitors - Failure', async () => {
  const wrapper = shallowMount(SafetyCheck, {
    stubs: ['router-link'],
    mocks: {
      $archive: {
        connection: {},
        connected: true,
        query: jest.fn().mockRejectedValue(),
      },
      $database: {
        connection: {},
        connected: true,
        query: jest.fn().mockRejectedValue([1, 2, 3]),
      },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
    },
  })
  await wrapper.vm.getOutstandingCompetitors()
  expect(wrapper.vm.$database.query).toHaveBeenCalledWith(expect.any(String), [
    12,
  ])
  expect(wrapper.vm.outstandingCompetitors).toEqual([])
  expect(wrapper.vm.$messages.addMessage).toHaveBeenCalledWith(
    'Problem Fetching Outstanding Competitors',
    'error'
  )
})

test('Filter Existing Punches', async () => {
  const wrapper = shallowMount(SafetyCheck, {
    stubs: ['router-link'],
    mocks: {
      $archive: {
        connection: {},
        connected: true,
        query: jest.fn().mockRejectedValue(),
      },
      $database: {
        connection: {},
        connected: true,
        query: jest.fn().mockResolvedValue([]),
      },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
    },
  })
  expect(
    await wrapper.vm.filterExistingPunches([
      { siid: 1, time: 1, controlCode: 1 },
      { siid: 3, time: 2, controlCode: 1 },
      { siid: 1, time: 3, controlCode: 1 },
    ])
  ).toEqual([
    { siid: 1, time: 1, controlCode: 1 },
    { siid: 3, time: 2, controlCode: 1 },
    { siid: 1, time: 3, controlCode: 1 },
  ])

  wrapper.vm.$database.query.mockResolvedValue([[1, 1, 1]])

  expect(
    await wrapper.vm.filterExistingPunches([
      { siid: 1, time: 1, controlCode: 1 },
      { siid: 3, time: 2, controlCode: 1 },
      { siid: 1, time: 3, controlCode: 1 },
    ])
  ).toEqual([
    { siid: 3, time: 2, controlCode: 1 },
    { siid: 1, time: 3, controlCode: 1 },
  ])

  expect(wrapper.vm.$database.query).toHaveBeenCalledWith(expect.any(String), [
    12,
  ])
})

test('Import Punches', async () => {
  const wrapper = shallowMount(SafetyCheck, {
    stubs: ['router-link'],
    mocks: {
      $archive: {
        connection: {},
        connected: true,
        query: jest.fn().mockRejectedValue(),
      },
      $database: {
        connection: {},
        connected: true,
        query: jest.fn().mockResolvedValue({ affectedRows: 3 }),
      },
      $electron: {
        remote: {
          dialog: {
            showOpenDialog: jest
              .fn()
              .mockResolvedValue({ canceled: false, filePaths: ['hi'] }),
          },
        },
      },
      $fs: { readFile: jest.fn() },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
    },
  })
  wrapper.setMethods({
    rawFileToPunchRecords: jest.fn().mockResolvedValue(),
    findCompetitorForPunches: jest.fn().mockResolvedValue(),
    filterExistingPunches: jest.fn().mockResolvedValue(),
    punchRecordToArray: jest.fn().mockResolvedValue([1, 2, 3]),
    getOutstandingCompetitors: jest.fn().mockResolvedValue(),
  })
  await wrapper.vm.importPunches()
  expect(wrapper.vm.rawFileToPunchRecords).toHaveBeenCalledTimes(1)
  expect(wrapper.vm.findCompetitorForPunches).toHaveBeenCalledTimes(1)
  expect(wrapper.vm.filterExistingPunches).toHaveBeenCalledTimes(1)
  expect(wrapper.vm.punchRecordToArray).toHaveBeenCalledTimes(1)
  expect(wrapper.vm.getOutstandingCompetitors).toHaveBeenCalledTimes(1)
  expect(
    wrapper.vm.$electron.remote.dialog.showOpenDialog
  ).toHaveBeenCalledTimes(1)
  expect(wrapper.vm.$messages.addMessage).toHaveBeenCalledWith(
    '3 Punches Imported'
  )
  expect(wrapper.vm.$fs.readFile).toHaveBeenCalledWith('hi', {
    encoding: 'utf8',
  })
  expect(
    wrapper.vm.$database.query
  ).toHaveBeenCalledWith(
    'INSERT INTO punches (competitor,controlCode,time,event,type) VALUES ?',
    [[1, 2, 3]]
  )
})

test('Import Punches - Cancelled', async () => {
  const wrapper = shallowMount(SafetyCheck, {
    stubs: ['router-link'],
    mocks: {
      $archive: {
        connection: {},
        connected: true,
        query: jest.fn().mockRejectedValue(),
      },
      $database: {
        connection: {},
        connected: true,
        query: jest.fn().mockResolvedValue({ affectedRows: 3 }),
      },
      $electron: {
        remote: {
          dialog: {
            showOpenDialog: jest
              .fn()
              .mockResolvedValue({ canceled: true, filePaths: ['hi'] }),
          },
        },
      },
      $fs: { readFile: jest.fn() },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
    },
  })
  wrapper.setMethods({
    rawFileToPunchRecords: jest.fn().mockResolvedValue(),
    findCompetitorForPunches: jest.fn().mockResolvedValue(),
    filterExistingPunches: jest.fn().mockResolvedValue(),
    punchRecordToArray: jest.fn().mockResolvedValue([1, 2, 3]),
    getOutstandingCompetitors: jest.fn().mockResolvedValue(),
  })
  await wrapper.vm.importPunches()
  expect(wrapper.vm.$messages.addMessage).toHaveBeenCalledWith(
    'Problem Importing Punches',
    'error'
  )
  expect(wrapper.vm.$fs.readFile).not.toHaveBeenCalledWith('hi', {
    encoding: 'utf8',
  })
})

test('Import Punches - No Punches', async () => {
  const wrapper = shallowMount(SafetyCheck, {
    stubs: ['router-link'],
    mocks: {
      $archive: {
        connection: {},
        connected: true,
        query: jest.fn().mockRejectedValue(),
      },
      $database: {
        connection: {},
        connected: true,
        query: jest.fn().mockResolvedValue({ affectedRows: 3 }),
      },
      $electron: {
        remote: {
          dialog: {
            showOpenDialog: jest
              .fn()
              .mockResolvedValue({ canceled: false, filePaths: ['hi'] }),
          },
        },
      },
      $fs: { readFile: jest.fn() },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
    },
  })
  wrapper.setMethods({
    rawFileToPunchRecords: jest.fn().mockResolvedValue(),
    findCompetitorForPunches: jest.fn().mockResolvedValue(),
    filterExistingPunches: jest.fn().mockResolvedValue(),
    punchRecordToArray: jest.fn().mockResolvedValue([]),
    getOutstandingCompetitors: jest.fn().mockResolvedValue(),
  })
  await wrapper.vm.importPunches()
  expect(wrapper.vm.rawFileToPunchRecords).toHaveBeenCalledTimes(1)
  expect(wrapper.vm.findCompetitorForPunches).toHaveBeenCalledTimes(1)
  expect(wrapper.vm.filterExistingPunches).toHaveBeenCalledTimes(1)
  expect(wrapper.vm.punchRecordToArray).toHaveBeenCalledTimes(1)
  expect(wrapper.vm.getOutstandingCompetitors).toHaveBeenCalledTimes(1)
  expect(
    wrapper.vm.$electron.remote.dialog.showOpenDialog
  ).toHaveBeenCalledTimes(1)
  expect(wrapper.vm.$messages.addMessage).toHaveBeenCalledWith(
    '0 Punches Imported'
  )
  expect(wrapper.vm.$fs.readFile).toHaveBeenCalledWith('hi', {
    encoding: 'utf8',
  })
})

test('Find Competitor For Punches', async () => {
  const wrapper = shallowMount(SafetyCheck, {
    stubs: ['router-link'],
    mocks: {
      $archive: {
        connection: {},
        connected: true,
        query: jest
          .fn()
          .mockResolvedValueOnce([{ name: 'Bob', membershipNumber: 3 }])
          .mockResolvedValueOnce([]),
      },
      $database: {
        connection: {},
        connected: true,
        query: jest.fn().mockResolvedValue(),
      },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
    },
  })

  wrapper.vm.$database.query = jest
    .fn()
    .mockResolvedValueOnce([{ id: 3 }])
    .mockResolvedValueOnce([])
    .mockResolvedValueOnce({ insertId: 2 })
    .mockResolvedValueOnce([])
    .mockResolvedValueOnce({ insertId: 1 })

  expect(
    await wrapper.vm.findCompetitorForPunches([
      { siid: 1 },
      { siid: 2 },
      { siid: 3 },
    ])
  ).toEqual([
    { siid: 1, competitor: 3 },
    { siid: 2, competitor: 2 },
    { siid: 3, competitor: 1 },
  ])

  expect(wrapper.vm.$archive.query).toHaveBeenCalledTimes(2)
  expect(wrapper.vm.$database.query).toHaveBeenCalledTimes(5)
  expect(wrapper.vm.$database.query).toHaveBeenCalledWith(
    'INSERT INTO competitors SET ?',
    {
      ageClass: '',
      club: undefined,
      downloaded: false,
      event: 12,
      membershipNumber: 3,
      name: 'Bob',
      siid: 2,
    }
  )
  expect(wrapper.vm.$database.query).toHaveBeenCalledWith(
    'INSERT INTO competitors SET ?',
    {
      downloaded: false,
      event: 12,
      name: 'Unknown',
      siid: 3,
    }
  )
})

test('Change Sort By', () => {
  const wrapper = shallowMount(SafetyCheck, {
    stubs: ['router-link'],
    mocks: {
      $database: {
        connection: {},
        connected: true,
        query: jest.fn().mockRejectedValue(),
      },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
    },
  })
  wrapper.setMethods({
    getOutstandingCompetitors: jest.fn(),
  })
  wrapper.vm.changeSortBy('name')
  expect(wrapper.vm.sortBy).toBe('name')
  expect(wrapper.vm.sortDirection).toBe('DESC')
  expect(wrapper.vm.getOutstandingCompetitors).toHaveBeenCalledTimes(1)
  wrapper.vm.changeSortBy('name')
  expect(wrapper.vm.sortBy).toBe('name')
  expect(wrapper.vm.sortDirection).toBe('ASC')
  expect(wrapper.vm.getOutstandingCompetitors).toHaveBeenCalledTimes(2)
  wrapper.vm.changeSortBy('siid')
  expect(wrapper.vm.sortBy).toBe('siid')
  expect(wrapper.vm.sortDirection).toBe('ASC')
  expect(wrapper.vm.getOutstandingCompetitors).toHaveBeenCalledTimes(3)
  wrapper.vm.changeSortBy('siid')
  expect(wrapper.vm.sortBy).toBe('siid')
  expect(wrapper.vm.sortDirection).toBe('DESC')
  expect(wrapper.vm.getOutstandingCompetitors).toHaveBeenCalledTimes(4)
})
