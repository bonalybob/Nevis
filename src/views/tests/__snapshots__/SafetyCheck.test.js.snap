// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Correctly 1`] = `
<main>
  <back-arrow-stub
    to="/events/undefined"
  />
   
  <h1
    class="mx-10 mb-1 px-1"
  >
    
    Safety Check
  
  </h1>
   
  <div
    class="mx-12 mb-3"
  >
    <button
      class="button"
    >
      Import Punches
    </button>
     
    <button
      class="button"
    >
      Refresh
    </button>
  </div>
   
  <div
    class="mx-12 my-shadow mx-12 mb-3 p-2"
  >
    <table
      class="w-full font-body"
    >
      <tr
        class="font-heading text-center hover:bg-blue-light"
      >
        <th>
          
          Name
          
          <up-down-arrows-stub />
        </th>
         
        <th>
          
          Control
          
          <up-down-arrows-stub />
        </th>
         
        <th>
          
          Time
          
          <up-down-arrows-stub
            active="true"
          />
        </th>
         
        <th
          class="hidden sm:table-cell"
        >
          
          Course
          
          <up-down-arrows-stub />
        </th>
         
        <th
          class="hidden md:table-cell"
        >
          
          Age Class
          
          <up-down-arrows-stub />
        </th>
      </tr>
       
      <tbody
        name="fade"
      >
        <tr
          class="text-center odd:bg-blue-lightest hover:bg-blue-light"
        >
          <td>
            Bob
          </td>
           
          <td>
            202
          </td>
           
          <td>
            01:34:15
          </td>
           
          <td
            class="hidden sm:table-cell"
          >
            Long
          </td>
           
          <td
            class="hidden md:table-cell"
          >
            
          </td>
        </tr>
        <tr
          class="text-center odd:bg-blue-lightest hover:bg-blue-light"
        >
          <td>
            James
          </td>
           
          <td>
            202
          </td>
           
          <td>
            01:15:55
          </td>
           
          <td
            class="hidden sm:table-cell"
          >
            Short
          </td>
           
          <td
            class="hidden md:table-cell"
          >
            
          </td>
        </tr>
        <tr
          class="text-center odd:bg-blue-lightest hover:bg-blue-light"
        >
          <td>
            Alfie
          </td>
           
          <td>
            201
          </td>
           
          <td>
            01:34:25
          </td>
           
          <td
            class="hidden sm:table-cell"
          >
            
          </td>
           
          <td
            class="hidden md:table-cell"
          >
            
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</main>
`;
